File: examples/typescript/case02.test.ts
import assert from "node:assert/strict";
import { test } from "node:test";
import { JobStore, commitCurrentWorker } from "./case02.ts";

test("stale generation cannot commit while current generation can", () => {
  const weak = new JobStore();
  weak.takeover(8); // B has not committed; still RUNNING.
  assert.equal(weak.storeResultWeak("A7"), true);
  assert.equal(weak.result, "A7");

  const gated = new JobStore();
  gated.takeover(8);
  assert.equal(gated.storeResultGated(7, "A7"), false);
  assert.equal(gated.result, null);
  assert.equal(commitCurrentWorker(gated, "B8"), true);
  assert.equal(gated.result, "B8");
});

File: examples/cpp/case02_test.cpp
#include <cassert>
#include "case02.cpp"

int main() {
    Case02JobStore weak;
    weak.takeover(8); // B has not committed; still RUNNING.
    assert(weak.store_result_weak("A7"));
    assert(weak.result == "A7");
    Case02JobStore gated;
    gated.takeover(8);
    assert(!gated.store_result_gated(7, "A7"));
    assert(gated.result.empty());
    assert(case02_commit_current_worker(gated, "B8"));
    assert(gated.result == "B8");
}
